File: app/macos/hyperwhisper/Managers/LocalAPI/LocalAPIErrors.swift
//
//  LocalAPIErrors.swift
//  hyperwhisper
//
//  Maps thrown errors from transcription/provider code into APIFailureEnvelope
//  responses, and provides small helpers for shaping JSON HTTPResponses with
//  the standard `{ok:true, ...}` / `{ok:false, error:{...}}` shape.
//

import Foundation
import FlyingFox

enum LocalAPIResponder {

    /// JSON encoder used for every API response. ISO-8601 dates so MCP
    /// clients don't have to know about Foundation's reference-date epoch.
    static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = [.withoutEscapingSlashes, .sortedKeys]
        e.dateEncodingStrategy = .iso8601
        return e
    }()

    static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }()

    /// Encode an `Encodable` payload into a successful HTTP 200 JSON response.
    static func ok<T: Encodable>(_ payload: T) -> HTTPResponse {
        do {
            let data = try encoder.encode(payload)
            return HTTPResponse(
                statusCode: .ok,
                headers: [.contentType: "application/json; charset=utf-8"],
                body: data
            )
        } catch {
            AppLogger.network.error("LocalAPI encoding failure · \(error.localizedDescription, privacy: .public)")
            // Fallback: minimal failure envelope so the client still gets JSON.
            return failure(code: .transcriptionFailed, message: "Failed to encode response", hint: nil)
        }
    }

    /// Standard `{ok:false, error:{...}}` response (HTTP 200 by design — see
    /// the rationale in the plan: MCP wrappers can't surface error text from
    /// an empty 500).
    static func failure(code: LocalAPIErrorCode, message: String, hint: String? = nil) -> HTTPResponse {
        let envelope = APIFailureEnvelope(code: code, message: message, hint: hint)
        do {
            let data = try encoder.encode(envelope)
            return HTTPResponse(
                statusCode: .ok,
                headers: [.contentType: "application/json; charset=utf-8"],
                body: data
            )
        } catch {
            // Truly defensive — if a 3-key envelope won't encode, return raw text.
            return HTTPResponse(
                statusCode: .internalServerError,
                headers: [.contentType: "text/plain; charset=utf-8"],
                body: Data("{\"ok\":false,\"error\":{\"code\":\"TRANSCRIPTION_FAILED\",\"message\":\"encoder broken\"}}".utf8)
            )
        }
    }

    /// Render a shared-core failure as an `HTTPResponse` (issue #289).
    ///
    /// The status code comes from Rust rather than from this file. That is the
    /// point of the envelope half of #289: Linux returned 404/413/503/408 for
    /// business outcomes the docs mandate 200 for, and it did so because each
    /// platform decided the status itself.
    ///
    /// The JSON is still encoded here, through the same `JSONEncoder` every
    /// other response uses, so key order and escaping do not change on the
    /// wire. `extraHeaders` carries the pieces that are not part of the
    /// envelope — `WWW-Authenticate` on a 401.
    static func response(
        for failure: HwLocalApiFailure,
        extraHeaders: [HTTPHeader: String] = [:]
    ) -> HTTPResponse {
        let envelope = APIFailureEnvelope(
            code: LocalAPIErrorCode(shared: failure.code),
            message: failure.message,
            hint: failure.hint
        )
        var headers: [HTTPHeader: String] = [.contentType: "application/json; charset=utf-8"]
        headers.merge(extraHeaders) { _, extra in extra }
        return HTTPResponse(
            statusCode: statusCode(failure.httpStatus),
            headers: headers,
            body: (try? encoder.encode(envelope)) ?? Data()
        )
    }

    /// The four statuses `hw_localapi` can ask for, as FlyingFox constants.
    ///
    /// `FailureKind` has exactly four cases and this switch covers all four;
    /// the default exists because the value crosses an FFI boundary as a
    /// `UInt16` and Swift cannot see that it is closed. A fifth status would
    /// be a contract change that needs this arm updated, not a 500.
    private static func statusCode(_ status: UInt16) -> HTTPStatusCode {
        switch status {
        case 200: return .ok
        case 400: return .badRequest
        case 401: return .unauthorized
        case 403: return .forbidden
        default: return .internalServerError
        }
    }

    /// Reserved for genuine protocol failures (malformed JSON body, etc.).
    /// Per the design, we keep these as HTTP 400 to distinguish from
    /// successful-but-unsuccessful business outcomes.
    static func badRequest(message: String, hint: String? = nil) -> HTTPResponse {
        let envelope = APIFailureEnvelope(code: .invalidRequest, message: message, hint: hint)
        let data = (try? encoder.encode(envelope)) ?? Data()
        return HTTPResponse(
            statusCode: .badRequest,
            headers: [.contentType: "application/json; charset=utf-8"],
            body: data
        )
    }

    /// Translate a thrown `TranscriptionError` into the corresponding
    /// `LocalAPIErrorCode` plus a human-readable message + hint pair.
    /// This is the single mapping point used by `/transcribe` so the
    /// caller never sees raw `TranscriptionError` text leaking through.
    ///
    /// THE TABLE ITSELF IS THE SHARED CORE'S (issue #356 item 4). This head
    /// used to hold one of three hand-maintained `(code, message, hint)` tables
    /// — Windows and the portable head had the other two, and they disagreed on
    /// wording, on hints, and in two places on the code. All this function does
    /// now is name the reason, fill the interpolation slots off the associated
    /// values, and pass in the one hint that has to name a macOS surface.
    ///
    /// The wording that comes back is macOS's on almost every row: this is the
    /// only head whose strings are under test
    /// (`HyperWhisperCloudEntitlementTests`), so the reconciliation was done
    /// toward it. `hw-localapi/src/transcription.rs` records the rows that went
    /// the other way.
    ///
    /// Not to be confused with `RustCoreMapping.mapTranscriptionError`
    /// (`RustRetry.swift`), which turns a Rust `HwTranscriptionError` into the
    /// `TranscriptionError` this function consumes, and is upstream of it.
    static func mapTranscriptionError(_ error: Error) -> (LocalAPIErrorCode, String, String?) {
        let (reason, params) = transcriptionFailure(for: error)
        let failure = localApiMapTranscriptionError(reason: reason, params: params)
        return (LocalAPIErrorCode(shared: failure.code), failure.message, failure.hint)
    }

    /// The shared reason for a thrown error, and the slots its message and hint
    /// interpolate.
    private static func transcriptionFailure(
        for error: Error
    ) -> (HwLocalApiTranscriptionFailureReason, HwLocalApiTranscriptionFailureParams) {
        guard let txError = error as? TranscriptionError else {
            return (.transcriptionFailed, slots(detail: error.localizedDescription))
        }
        switch txError {
        case .modelNotDownloaded:
            return (.modelNotInstalled, slots())
        case .modelProtected:
            return (.modelProtected, slots())
        case .apiKeyMissing(let provider):
            return (.apiKeyMissing, slots(provider: provider, hint: "Add the API key in Settings → API Keys."))
        case .unauthorized(let provider, let statusCode):
            // A HyperWhisper Cloud 403 is the abuse guard, not a credential
            // fault: `transcribe`, `post-process`, `usage` and `assistant`
            // all answer 403 only for "Your IP has been temporarily blocked
            // due to abuse". Reporting MISSING_API_KEY there tells an API
            // client to rotate a key that is in fact valid, so this maps to
            // the temporary code instead. It is its own shared reason, and a
            // test pins that it stays RATE_LIMITED with no Settings hint.
            if provider == "HyperWhisper Cloud", statusCode == 403 {
                return (.cloudRequestForbidden, slots(provider: provider))
            }
            let settingsDestination = provider == "HyperWhisper Cloud"
                ? "Settings → HyperWhisper Cloud"
                : "Settings → API Keys"
            // `statusCode` is deliberately not passed: the crate renders an
            // HTTP status only for `.providerServerError`, and a 401 in the
            // message would say nothing an "invalid or expired" key does not.
            return (.apiKeyInvalid, slots(provider: provider, hint: "Update the API key in \(settingsDestination)."))
        case .audioFileNotFound:
            return (.audioFileNotFound, slots())
        case .invalidAudioFormat, .audioConversionFailed:
            return (.audioDecodeFailed, slots())
        case .audioFileTooLarge(_, let limit, let providerName):
            return (.audioFileTooLarge, slots(provider: providerName, limitBytes: UInt64(exactly: limit)))
        case .rateLimited(let retryAfter):
            // This head threw the `retryAfter` away and always said "Try again
            // in a moment."; the shared row renders "Retry after N seconds."
            // when the provider told us how long, which is Windows's hint.
            return (.rateLimited, slots(retryAfterSeconds: retryAfter.flatMap { UInt32(exactly: $0) }))
        case .timeout:
            return (.timeout, slots())
        case .providerNotAvailable(let provider, let reason):
            return (.engineUnavailable, slots(provider: provider, detail: reason))
        case .transientNetwork(let details):
            return (.networkUnavailable, slots(detail: details))
        case .invalidResponse(let details):
            return (.invalidProviderResponse, slots(detail: details))
        case .invalidRequest:
            return (.invalidRequest, slots())
        case .serverError(let statusCode, let message):
            return (.providerServerError, slots(detail: message, httpStatus: UInt16(exactly: statusCode)))
        case .noSpeechDetected:
            return (.noSpeechDetected, slots())
        case .cloudAccountRequired:
            // Reuses MISSING_API_KEY rather than adding a code: LocalAPIErrorCode
            // is a closed enum decoded by cross-platform clients, and this IS a
            // missing-credential condition. Without an arm it fell through to
            // the generic TRANSCRIPTION_FAILED, which is exactly the leak this
            // function exists to prevent.
            return (.cloudAccountRequired, slots(hint: "Add your account key in Settings → HyperWhisper Cloud."))
        case .localSpeechModelEvicted(let model):
            // ENGINE_UNAVAILABLE, not the generic TRANSCRIPTION_FAILED the
            // `default:` below would give it: the engine genuinely was not
            // resident, and a Local API client can act on that (retry later)
            // in a way it cannot act on "transcription failed".
            return (.localModelEvicted, slots(model: model))
        case .quotaExceeded(let provider, _):
            // No arm before this change: both of these reached the wire as the
            // generic TRANSCRIPTION_FAILED plus a localized
            // `localizedDescription`, and RATE_LIMITED is the code the closed
            // set has for exactly this. Windows already answered RATE_LIMITED.
            // The associated `message` is dropped: it is the provider's own
            // billing prose, and the shared row carries the action in its hint.
            return (.quotaExceeded, slots(provider: provider))
        case .insufficientCredits:
            // Same row. The `remaining`/`required` counts are dropped for the
            // same reason the provider message is: the shared row's hint says
            // what to do, and no other head has the numbers to render.
            return (.quotaExceeded, slots())
        default:
            // Four cases stay here on purpose, and each would need a wording
            // decision no head has made:
            //
            // * `.maxRetriesExceeded` and `.streamingInterrupted` — the
            //   underlying cause (network, rate limit, provider error) is gone
            //   by the time either is thrown, so any specific code would be a
            //   guess.
            // * `.busy` — the app is mid-transcription. ENGINE_UNAVAILABLE's
            //   hint says "pick a different engine", which is wrong advice.
            // * `.localRuntimeUnavailable` — names the local LLM used for
            //   POST-processing, not a transcription engine, and its own doc
            //   comment says the raw `reason` is for logs and must not be
            //   shown. ENGINE_UNAVAILABLE would put the wrong noun on the one
            //   route (`/post-process`) that produces it.
            //
            // The generic row passes `localizedDescription` through verbatim,
            // which is exactly what this arm always did.
            return (.transcriptionFailed, slots(detail: error.localizedDescription))
        }
    }

    /// The shared params record, with every slot defaulted to absent.
    ///
    /// The generated initializer takes all seven, so calling it inline at
    /// twenty sites would bury the two or three values that matter in five
    /// `nil`s.
    private static func slots(
        provider: String? = nil,
        detail: String? = nil,
        model: String? = nil,
        limitBytes: UInt64? = nil,
        httpStatus: UInt16? = nil,
        retryAfterSeconds: UInt32? = nil,
        hint: String? = nil
    ) -> HwLocalApiTranscriptionFailureParams {
        HwLocalApiTranscriptionFailureParams(
            provider: provider,
            detail: detail,
            model: model,
            limitBytes: limitBytes,
            httpStatus: httpStatus,
            retryAfterSeconds: retryAfterSeconds,
            hint: hint
        )
    }
}

File: app/windows/HyperWhisper/Services/LocalApi/LocalApiErrors.cs
using System.Text.Json;
using HyperWhisper.Models;
using Microsoft.AspNetCore.Http;
using uniffi.hyperwhisper_core;

namespace HyperWhisper.Services.LocalApi;

/// <summary>
/// Shapes successful and failure JSON responses for the Local API. Mirrors
/// the macOS LocalAPIResponder so wire shapes stay 1:1 across platforms.
/// </summary>
internal static class LocalApiResponder
{
    public static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = false,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
    };

    /// <summary>200 JSON with the given payload.</summary>
    public static IResult Ok<T>(T payload)
    {
        return Results.Json(payload, JsonOptions, contentType: "application/json; charset=utf-8", statusCode: 200);
    }

    /// <summary>
    /// `{ok:false, error:{...}}` business-failure envelope returned with HTTP
    /// 200 by design — MCP wrappers can't surface error text from an empty 500.
    /// </summary>
    public static IResult Failure(string code, string message, string? hint = null)
    {
        var envelope = new ApiFailureEnvelope
        {
            Error = new ApiError { Code = code, Message = message, Hint = hint }
        };
        return Results.Json(envelope, JsonOptions, contentType: "application/json; charset=utf-8", statusCode: 200);
    }

    /// <summary>Genuine protocol failure (malformed JSON, bad path, etc.).</summary>
    /// <summary>
    /// A failure <c>hw-localapi</c> already decided — status, code, message and
    /// hint all come from the crate (issue #356).
    /// </summary>
    /// <remarks>
    /// The crate also encodes the whole envelope as JSON, but this head writes
    /// its own <see cref="ApiFailureEnvelope"/> so the response goes out under
    /// the same <see cref="JsonOptions"/> as every other response on this
    /// server. The parts that must not drift — which of the closed fourteen,
    /// which status, and the exact wording — are the crate's.
    /// </remarks>
    public static IResult Shared(HwLocalApiFailure failure)
    {
        ArgumentNullException.ThrowIfNull(failure);
        var code = HyperwhisperCoreMethods.LocalApiErrorCodeWireValue(failure.code);
        return failure.httpStatus == 400
            ? BadRequestWithCode(code, failure.message, failure.hint)
            : Failure(code, failure.message, failure.hint);
    }

    private static IResult BadRequestWithCode(string code, string message, string? hint)
    {
        var envelope = new ApiFailureEnvelope
        {
            Error = new ApiError { Code = code, Message = message, Hint = hint }
        };
        return Results.Json(envelope, JsonOptions, contentType: "application/json; charset=utf-8", statusCode: 400);
    }

    public static IResult BadRequest(string message, string? hint = null)
    {
        var envelope = new ApiFailureEnvelope
        {
            Error = new ApiError { Code = LocalApiErrorCode.InvalidRequest, Message = message, Hint = hint }
        };
        return Results.Json(envelope, JsonOptions, contentType: "application/json; charset=utf-8", statusCode: 400);
    }

    /// <summary>
    /// Translate a <see cref="TranscriptionException"/> into the wire-level
    /// (code, message, hint) tuple the Failure responder expects. Shared by
    /// `/transcribe` (Phase 2) and `/post-process` (Phase 3).
    /// </summary>
    /// <remarks>
    /// THE TABLE IS THE SHARED CORE'S (issue #356 item 4). This method now does
    /// one job: turn a <see cref="TranscriptionErrorCode"/> into one of the
    /// crate's reasons, fill the interpolation slots off the exception, and
    /// hand this head's own API-key hint to the three rows whose hint has to
    /// name a product surface. The code, the message and every other hint come
    /// back from <c>hw-localapi</c>, so they cannot drift from macOS's and the
    /// portable head's again.
    ///
    /// Every row is still a business failure at HTTP 200, which is what
    /// <see cref="Failure"/> writes and what this method's callers already did.
    ///
    /// TWO CODES CHANGE ON THIS HEAD, deliberately. There was no arm for
    /// <see cref="TranscriptionErrorCode.CloudAccountRequired"/> or
    /// <see cref="TranscriptionErrorCode.NoSpeechDetected"/>, so both fell to
    /// the `_` arm and answered <c>TRANSCRIPTION_FAILED</c>; macOS answers
    /// <c>MISSING_API_KEY</c> for the first. That is exactly the divergence
    /// item 4 exists to close, so this head follows macOS.
    ///
    /// The messages change on almost every row, because they used to come from
    /// <see cref="TranscriptionException.GetUserMessage"/> — which is written
    /// for a WPF toast, mixes the fix into the sentence, and is *localized* for
    /// <see cref="TranscriptionErrorCode.NoSpeechDetected"/>. The Local API
    /// wire is not localized and it carries the fix in a separate `hint` field.
    /// </remarks>
    public static (string code, string message, string? hint) MapTranscriptionException(TranscriptionException ex)
    {
        ArgumentNullException.ThrowIfNull(ex);
        var reason = ReasonFor(ex.Code);
        var failure = HyperwhisperCoreMethods.LocalApiMapTranscriptionError(
            reason,
            new HwLocalApiTranscriptionFailureParams(
                @provider: ex.ProviderName,
                // `GetUserMessage()` for a code with no arm of its own is just
                // `Message`, and the generic row is the only one that renders
                // the detail as the whole message — so the fallback text this
                // head has always sent survives.
                @detail: ex.GetUserMessage(),
                @model: null,
                @limitBytes: null,
                // Filled because this head has it. No reason this head can
                // reach reads it today — the crate renders it only for
                // `ProviderServerError`, which is macOS's `.serverError` and
                // has no `TranscriptionErrorCode` equivalent — so this is the
                // slot being kept honest, not a rendered value.
                @httpStatus: ex.HttpStatusCode is int status and >= 0 and <= ushort.MaxValue ? (ushort)status : null,
                @retryAfterSeconds: ex.RetryAfterSeconds is int seconds and >= 0 ? (uint)seconds : null,
                @hint: PlatformHint(reason)));
        return (HyperwhisperCoreMethods.LocalApiErrorCodeWireValue(failure.code), failure.message, failure.hint);
    }

    /// <summary>
    /// This head's <see cref="TranscriptionErrorCode"/> as one of the crate's
    /// reasons.
    /// </summary>
    /// <remarks>
    /// <see cref="TranscriptionErrorCode.Unknown"/> and anything added later
    /// land on the generic row, which is where the old `_` arm put them.
    /// </remarks>
    private static HwLocalApiTranscriptionFailureReason ReasonFor(TranscriptionErrorCode code) => code switch
    {
        TranscriptionErrorCode.ModelNotLoaded => HwLocalApiTranscriptionFailureReason.ModelNotInstalled,
        TranscriptionErrorCode.OnnxModelFileMissing => HwLocalApiTranscriptionFailureReason.ModelFilesMissing,
        TranscriptionErrorCode.ApiKeyMissing => HwLocalApiTranscriptionFailureReason.ApiKeyMissing,
        TranscriptionErrorCode.Unauthorized => HwLocalApiTranscriptionFailureReason.ApiKeyInvalid,
        TranscriptionErrorCode.CloudAccountRequired => HwLocalApiTranscriptionFailureReason.CloudAccountRequired,
        TranscriptionErrorCode.AudioFileNotFound => HwLocalApiTranscriptionFailureReason.AudioFileNotFound,
        TranscriptionErrorCode.UnsupportedFormat => HwLocalApiTranscriptionFailureReason.AudioDecodeFailed,
        TranscriptionErrorCode.FileTooLarge => HwLocalApiTranscriptionFailureReason.AudioFileTooLarge,
        TranscriptionErrorCode.InvalidRequest => HwLocalApiTranscriptionFailureReason.InvalidRequest,
        TranscriptionErrorCode.RateLimited => HwLocalApiTranscriptionFailureReason.RateLimited,
        TranscriptionErrorCode.QuotaExceeded => HwLocalApiTranscriptionFailureReason.QuotaExceeded,
        TranscriptionErrorCode.NetworkError => HwLocalApiTranscriptionFailureReason.NetworkUnavailable,
        TranscriptionErrorCode.ProviderUnavailable => HwLocalApiTranscriptionFailureReason.EngineUnavailable,
        TranscriptionErrorCode.DaemonStartFailed => HwLocalApiTranscriptionFailureReason.EngineStartFailed,
        TranscriptionErrorCode.DaemonCrashed => HwLocalApiTranscriptionFailureReason.EngineCrashed,
        TranscriptionErrorCode.DaemonTimeout => HwLocalApiTranscriptionFailureReason.EngineTimeout,
        TranscriptionErrorCode.NoSpeechDetected => HwLocalApiTranscriptionFailureReason.NoSpeechDetected,
        TranscriptionErrorCode.Cancelled => HwLocalApiTranscriptionFailureReason.Cancelled,
        _ => HwLocalApiTranscriptionFailureReason.TranscriptionFailed,
    };

    /// <summary>
    /// The hint for the three rows the crate deliberately leaves to the head,
    /// because they have to name a product surface: this app puts BYOK keys in
    /// the Model Library's API keys manager and the HyperWhisper Cloud account
    /// key in Settings → HyperWhisper Cloud, where macOS says
    /// `Settings → API Keys`.
    /// </summary>
    /// <remarks>
    /// The crate ignores this slot on the other twenty-two reasons — a head
    /// that could override any hint would put the wording back where item 4
    /// found it — so returning <c>null</c> for them is the honest answer and
    /// not a lost hint.
    /// </remarks>
    private static string? PlatformHint(HwLocalApiTranscriptionFailureReason reason) => reason switch
    {
        HwLocalApiTranscriptionFailureReason.ApiKeyMissing =>
            "Add the API key in the Model Library API keys manager.",
        HwLocalApiTranscriptionFailureReason.ApiKeyInvalid =>
            "Check that the API key in the Model Library API keys manager is valid.",
        // NOT the API keys manager: the account key is entered on the Cloud
        // settings page, which is the same split `MainViewModel` makes when it
        // routes this error's toast (`settingsSection: "Cloud"`).
        HwLocalApiTranscriptionFailureReason.CloudAccountRequired =>
            "Add your account key in Settings → HyperWhisper Cloud.",
        _ => null,
    };

    /// <summary>Missing or invalid bearer token.</summary>
    public static IResult Unauthorized()
    {
        var envelope = new ApiFailureEnvelope
        {
            Error = new ApiError
            {
                Code = LocalApiErrorCode.InvalidRequest,
                Message = "Missing or invalid bearer token",
                Hint = @"Send Authorization: Bearer <token>; the token lives in %LOCALAPPDATA%\HyperWhisper\local-api.json."
            }
        };
        return Results.Json(envelope, JsonOptions, contentType: "application/json; charset=utf-8", statusCode: 401);
    }
}
